### PR TITLE
feat(RTE): add text styles dropdown

### PR DESCRIPTION
**Pull request by celine-s**
Ticket on clickup: https://app.clickup.com/t/1rp6wz4

### CODE REVIEW

File: src/components/ActionMenu/Aria/helper.tsx [hunk 1]
```diff
@@ -15,7 +15,7 @@ export const mapToAriaProps = (ariaLabel: string, menuBlocks: MenuBlock[] | Acti
     children: menuBlocks.map(({ id, menuItems, ariaLabel: sectionAriaLabel }) => (
         <Section title={id} key={id} aria-label={sectionAriaLabel}>
             {menuItems.map((menuItem) => (
-                <Item key={menuItem.id}>{menuItem.title}</Item>
+                <Item key={menuItem.id}>{typeof menuItem.title === "string" ? menuItem.title : "title"}</Item>
```
Comment (NadiaPosch) [inline src/components/ActionMenu/Aria/helper.tsx:18]: You can revert this change since it's not needed anymore due to the new `useSelect` logic

File: src/components/RichTextEditor/BlockStyleDropdown.tsx [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+                    send({
+                        type: "BLOCK_TYPE_SELECTED",
+                        data: {
+                            type: id as BlockStyleTypes,
+                            active: !!blockTypeIsActive,
+                            editor,
+                        },
+                    });
+                }}
+                selectedKey={blockTypeId || "Body Text"}
```
Comment (NadiaPosch) [inline src/components/RichTextEditor/BlockStyleDropdown.tsx:67]:
```suggestion
                selectedKey={blockTypeId || BlockStyleTypes.Paragraph}
```

File: src/components/RichTextEditor/TextStyleDropdown/ListBox.tsx [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    );
+
+    return (
+        <li
+            {...optionProps}
+            ref={ref}
+            className={`tw-border-b tw-p-3 tw-border-black-30 tw-text-sm tw-outline-none tw-cursor-default tw-flex tw-items-center tw-justify-between tw-truncate ${
+                isFocused || isSelected ? "tw-bg-black-100 tw-text-white " : ""
+            }`}
+            data-test-id="menu-item-dropdown"
```
Comment (NadiaPosch) [inline src/components/RichTextEditor/TextStyleDropdown/ListBox.tsx:79]: I'd maybe rename this to `data-test-id="select-option"` in order not to confuse it with the menu item we have in the actual Dropdown component

File: src/components/RichTextEditor/TextStyleDropdown/Select.tsx [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+                {label}
+            </div>
+            <HiddenSelect state={state} triggerRef={ref} label={label} name={name} />
+            <button
+                {...mergeProps(buttonProps, focusProps)}
+                ref={ref}
+                className={`tw-p-2 tw-pl-3 tw-py-1 tw-relative tw-inline-flex tw-flex-row tw-items-center tw-justify-between tw-rounded tw-overflow-hidden tw-w-32 tw-cursor-default tw-shadow-sm tw-border tw-outline-none hover:tw-border-black-100 ${
+                    isFocusVisible ? "tw-border-black-50" : "tw-border-black-30"
+                } `}
+                data-test-id="text-select-dropdown"
```
Comment (NadiaPosch) [inline src/components/RichTextEditor/TextStyleDropdown/Select.tsx:38]: I'd maybe rename this to `data-test-id="select-trigger"` since it's a button and not the actual dropdown

File: src/components/RichTextEditor/RichTextEditor.stories.tsx [hunk 5]
```diff
@@ -108,3 +117,19 @@ export const RichTextEditorFlex: Story<RichTextEditorProps> = (args: RichTextEdi
...
+WithCustomTextStyle.args = {
+    readonly: false,
+    value: JSON.stringify([
+        ...value,
+        {
+            type: BlockStyleTypes.Paragraph,
+            children: [createLinkNode("https://git-scm.com/downloads", "Link")],
+        },
+    ]),
+    textStyles: [{ type: BlockStyleTypes.H3, className: "tw-text-xxl" }],
```
Comment (NadiaPosch) [inline src/components/RichTextEditor/RichTextEditor.stories.tsx:133]: To make the story a bit more explicit, I'd only display some headings and add custom styles for each in order to have a good working example for how to pass in textStyles, e.g.:
```suggestion
    value: JSON.stringify([
        { type: BlockStyleTypes.H1, children: [{ text: headingTitle[BlockStyleTypes.H1] }] },
        { type: BlockStyleTypes.H2, children: [{ text: headingTitle[BlockStyleTypes.H2] }] },
        { type: BlockStyleTypes.H3, children: [{ text: headingTitle[BlockStyleTypes.H3] }] },
        { type: BlockStyleTypes.H4, children: [{ text: headingTitle[BlockStyleTypes.H4] }] },
        { type: BlockStyleTypes.Custom01, children: [{ text: headingTitle[BlockStyleTypes.Custom01] }] },
        { type: BlockStyleTypes.Custom02, children: [{ text: headingTitle[BlockStyleTypes.Custom02] }] },
    ]),
    textStyles: [
        { type: BlockStyleTypes.H1, className: "tw-text-7xl tw-font-bold tw-text-green-80" },
        { type: BlockStyleTypes.H2, className: "tw-text-5xl tw-font-bold tw-text-violet-60" },
        { type: BlockStyleTypes.H3, className: "tw-text-3xl tw-font-bold tw-text-yellow-70" },
        { type: BlockStyleTypes.H4, className: "tw-text-xl tw-text-red-50" },
        { type: BlockStyleTypes.Custom01, className: "tw-font-mono tw-italic tw-text-black-80" },
        { type: BlockStyleTypes.Custom02, className: "tw-underline tw-text-black-80" },
    ],
```

File: src/components/RichTextEditor/RichTextEditor.tsx [hunk 6]
```diff
@@ -22,20 +23,34 @@ import { Toolbar } from "./Toolbar";
 import { clearEditor } from "./utils/editor/clear";
 import { parseRawValue } from "./utils/editor/parseRawContent";
 
+export type TextStylesType = {
```
Comment (NadiaPosch) [inline src/components/RichTextEditor/RichTextEditor.tsx:26]: I'd maybe rename this to just `TextStyles` because it includes not only the types but also the classNames

File: src/components/RichTextEditor/TextStyleDropdown/ListBox.tsx [hunk 7]
```diff
@@ -0,0 +1,84 @@
+/* eslint-disable @typescript-eslint/no-use-before-define */
+import { AriaListBoxOptions, useListBox, useListBoxSection, useOption } from "@react-aria/listbox";
+import { ListState } from "@react-stately/list";
+import type { Node } from "@react-types/shared";
+import * as React from "react";
```
Comment (NadiaPosch) [inline src/components/RichTextEditor/TextStyleDropdown/ListBox.tsx:5]:
```suggestion
import React from "react";
```

File: src/components/RichTextEditor/TextStyleDropdown/ListBox.tsx [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+interface ListBoxProps extends AriaListBoxOptions<unknown> {
+    listBoxRef?: React.RefObject<HTMLUListElement>;
```
Comment (NadiaPosch) [inline src/components/RichTextEditor/TextStyleDropdown/ListBox.tsx:8]: I'd suggest you import the RefObject directly
```suggestion
    listBoxRef?: RefObject<HTMLUListElement>;
```

File: src/components/RichTextEditor/TextStyleDropdown/ListBox.tsx [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    state: ListState<unknown>;
+}
+
+interface OptionProps {
+    item: Node<unknown>;
+    state: ListState<unknown>;
+}
+
+export function ListBox(props: ListBoxProps) {
+    const ref = React.useRef<HTMLUListElement>(null);
```
Comment (NadiaPosch) [inline src/components/RichTextEditor/TextStyleDropdown/ListBox.tsx:23]: same here

File: src/components/RichTextEditor/TextStyleDropdown/ListBox.tsx [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+                    <ListBoxSection key={item.key} section={item} state={state} />
+                ) : (
+                    <Option key={item.key} item={item} state={state} />
+                ),
+            )}
+        </ul>
+    );
+}
+
+function ListBoxSection({ section, state }: SectionProps) {
```
Comment (NadiaPosch) [inline src/components/RichTextEditor/TextStyleDropdown/ListBox.tsx:40]: Please move this to a separate file

File: src/components/RichTextEditor/TextStyleDropdown/ListBox.tsx [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+            <ul {...groupProps}>
+                {[...section.childNodes].map((node) => (
+                    <Option key={node.key} item={node} state={state} />
+                ))}
+            </ul>
+        </li>
+    );
+}
+
+function Option({ item, state }: OptionProps) {
```
Comment (NadiaPosch) [inline src/components/RichTextEditor/TextStyleDropdown/ListBox.tsx:62]: Please move this to a separate file

File: src/components/RichTextEditor/TextStyleDropdown/Popover.tsx [hunk 12]
```diff
@@ -0,0 +1,38 @@
+import { FocusScope } from "@react-aria/focus";
+import { DismissButton, useOverlay } from "@react-aria/overlays";
+import * as React from "react";
+
+interface PopoverProps {
+    popoverRef?: React.RefObject<HTMLDivElement>;
```
Comment (NadiaPosch) [inline src/components/RichTextEditor/TextStyleDropdown/Popover.tsx:6]: same here about the imports

File: src/components/RichTextEditor/TextStyleDropdown/Select.tsx [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+
+export function Select<T extends object>(props: AriaSelectProps<T>) {
+    const state = useSelectState(props);
+
+    const ref = useRef(null);
+    const { labelProps, triggerProps, valueProps, menuProps } = useSelect(props, state, ref);
+
+    const { buttonProps } = useButton(triggerProps, ref);
+
+    const { focusProps, isFocusVisible } = useFocusRing();
```
Comment (NadiaPosch) [inline src/components/RichTextEditor/TextStyleDropdown/Select.tsx:21]:
```suggestion
    const ref = useRef(null);
    const { labelProps, triggerProps, valueProps, menuProps } = useSelect(props, state, ref);
    const { buttonProps } = useButton(triggerProps, ref);
    const { focusProps, isFocusVisible } = useFocusRing();
```

File: src/components/RichTextEditor/TextStyleDropdown/Select.tsx [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+                className={`tw-p-2 tw-pl-3 tw-py-1 tw-relative tw-inline-flex tw-flex-row tw-items-center tw-justify-between tw-rounded tw-overflow-hidden tw-w-32 tw-cursor-default tw-shadow-sm tw-border tw-outline-none hover:tw-border-black-100 ${
+                    isFocusVisible ? "tw-border-black-50" : "tw-border-black-30"
+                } `}
+                data-test-id="text-select-dropdown"
+            >
+                <span
+                    {...valueProps}
+                    className={`tw-text-md ${selectedItem ? "tw-text-black-100" : "tw-text-black-70"}`}
+                >
+                    {selectedItem ? selectedItem.textValue : placeholder}
```
Comment (NadiaPosch) [inline src/components/RichTextEditor/TextStyleDropdown/Select.tsx:44]:
```suggestion
                    {selectedItem?.textValue || placeholder}
```

File: src/components/RichTextEditor/Toolbar.tsx [hunk 15]
```diff
@@ -16,13 +16,18 @@ import { IconSize } from "@foundation/Icon/IconSize";
...
 import { useEditorSelection } from "./hooks/useEditorSelection";
 import { InlineStyleButton } from "./InlineStyleButton";
 import { BlockStyleTypes, TextAlignTypes } from "./renderer/renderBlockStyles";
 import { InlineStyles } from "./renderer/renderInlineStyles";
 
-export const Toolbar: FC = () => {
+type ToolbarProps = {
+    textStyles?: TextStylesType[];
+};
+export const Toolbar: FC<ToolbarProps> = ({ textStyles }) => {
```
Comment (NadiaPosch) [inline src/components/RichTextEditor/Toolbar.tsx:30]:
```suggestion
type ToolbarProps = {
    textStyles?: TextStylesType[];
};

export const Toolbar: FC<ToolbarProps> = ({ textStyles }) => {
```

File: src/components/RichTextEditor/renderer/renderBlockStyles.tsx [hunk 16]
```diff
@@ -27,15 +35,31 @@ export const textAlignClassMap: Record<TextAlignTypes, string> = {
     [TextAlignTypes.AlignRight]: "tw-text-right",
 };
 
-export const renderBlockStyles = (props: RenderElementProps): JSX.Element => {
+export const headingTypeClassname = {
```
Comment (NadiaPosch) [inline src/components/RichTextEditor/renderer/renderBlockStyles.tsx:38]: I'd maybe rename this to `textStyleClassname` because it includes not only headings but also the custom and paragraph styles

File: src/components/RichTextEditor/renderer/renderBlockStyles.tsx [hunk 17]
```diff
@@ -27,15 +35,31 @@ export const textAlignClassMap: Record<TextAlignTypes, string> = {
...
     switch (props.element.type) {
         case BlockStyleTypes.Paragraph:
             const textAlign = props.element.properties?.textAlign;
             return (
-                <p {...props.attributes} className={textAlign ? textAlignClassMap[textAlign] : ""}>
+                <p
+                    {...props.attributes}
+                    className={`${getTextStyles(BlockStyleTypes.Paragraph, textStyles)} ${
+                        textAlign ? textAlignClassMap[textAlign] : ""
+                    }`}
```
Comment (NadiaPosch) [inline src/components/RichTextEditor/renderer/renderBlockStyles.tsx:57]: You could use the `merge` method from `@utilities/merge` here

File: src/components/RichTextEditor/state/toolbar/actions.ts [hunk 18]
```diff
@@ -29,15 +29,22 @@ export const updateInlineStyle = (_: ToolbarContext, { data }: DoneInvokeEvent<T
 
 const toggleBlock = (active: boolean, editor: Editor, type: BlockStyleTypes, textAlign?: TextAlignTypes) => {
     const isList = [BlockStyleTypes.OrderedList, BlockStyleTypes.UnorderedList].includes(type);
-
+    const isHeadingOrText = [
+        BlockStyleTypes.Paragraph,
+        BlockStyleTypes.H1,
+        BlockStyleTypes.H2,
+        BlockStyleTypes.H3,
```
Comment (NadiaPosch) [inline src/components/RichTextEditor/state/toolbar/actions.ts:36]:
```suggestion
        BlockStyleTypes.H3,
        BlockStyleTypes.H4,
```

File: src/components/RichTextEditor/RichTextEditor.stories.tsx (hunk 5 above)
Comment (celine-s) [reply src/components/RichTextEditor/RichTextEditor.stories.tsx:133]: nice!

File: src/components/ActionMenu/Aria/helper.tsx [hunk 19]
```diff
@@ -1,7 +1,7 @@
 import { Item, Section } from "@react-stately/collections";
 import React, { ReactNode } from "react";
-import { ActionMenuBlock, ActionMenuItemType } from "../ActionMenu/ActionMenu";
 import { MenuBlock, MenuItemType } from "../../Dropdown/SelectMenu/SelectMenu";
+import { ActionMenuBlock, ActionMenuItemType } from "../ActionMenu/ActionMenu";
```
Comment (NadiaPosch) [inline src/components/ActionMenu/Aria/helper.tsx:4]: I'd revert this change because you didn't actually change something in the file (you can apply "save without formatting" so it doesn't reorder the imports)

File: src/components/RichTextEditor/BlockStyleDropdown.tsx [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+        setBlockTypeId(nodeType);
+    }, [nodeType]);
+
+    const [, send] = useActor(machineRef);
+
+    const getMenuItemStyle = (identifier: TextStyles["type"]) =>
+        [getTextStyles(identifier, textStyles), "tw-truncate"].join(" ");
+
+    return (
+        <div>
```
Comment (NadiaPosch) [inline src/components/RichTextEditor/BlockStyleDropdown.tsx:54]: You can remove the div

File: src/components/RichTextEditor/RichTextEditor.spec.tsx [hunk 21]
```diff
@@ -12,6 +12,9 @@ const RICH_TEXT_EDITOR = "[data-test-id=rich-text-editor]";
 const TOOLBAR = "[data-test-id=toolbar]";
 const PLACEHOLDER = "My placeholder";
 
+const TEXT_STYLE_DROPDOWN_ID = "[data-test-id=select-trigger]";
+const MENU_ITEM_DROPDOWN_ID = "[data-test-id=select-option]";
```
Comment (NadiaPosch) [inline src/components/RichTextEditor/RichTextEditor.spec.tsx:16]: Maybe also rename the variables to `SELECT_TRIGGER_ID` and `SELECT_OPTION_ID`

**Review by triggertoo — APPROVED** [review]
💯

**Comment by triggertoo** [comment]
@celine-s There seems to be an issue i didn't notice yet, but the toolbar jumps to the top left after selecting a style.

**Comment by Walkeezy** [comment]
Rebuild all the changes from this PR inside of #612, so I'm closing this one. :)
